qutebrowser/mainwindow/prompt.py

```diff
@@ -467,6 +467,31 @@ def prompt_yank(self, sel=False):
         utils.set_clipboard(question.url, sel)
         message.info("Yanked to {}: {}".format(target, question.url))
 
+    @cmdutils.register(
+        instance='prompt-container', scope='window',
+        modes=[usertypes.KeyMode.prompt])
+    def prompt_fileselect_external(self):
+        """Choose a location using a configured external picker."""
+        if not (self._prompt and isinstance(self._prompt, FilenamePrompt)):
+            raise cmdutils.CommandError(
+                f"Can only launch external fileselect for FilenamePrompt, not {self._prompt}"
+            )
+        picker = config.val.fileselect.folder.command
+        if not picker:
+            raise cmdutils.CommandError(
+                "No external location picker (fileselect.folder.command) configured."
+            )
+        # TODO fix cyclic import
+        from qutebrowser.browser import shared
```

Comment:
The cycle is:

- `browser.browsertab`
- `misc.sessions` (for `sessions.session_manager.save_autosave()` in `_on_load_finished`)
- `mainwindow.mainwindow` (to create new windows from sessions)
- `mainwindow.prompt` (to add the prompt container)
- `browser.shared` (new, for file selection)
- `misc.guiprocess` (to spawn the file selector)
- `api.apitypes` (for `apitypes.Tab` type annotation)
- `browser.browsertab` (`apitypes.Tab` is an alias to `browsertab.Tab`)

No obvious point to break the cycle right now.
This solution is fine for now, I suppose. It's a bigger problem all across the code base, because it's not something I paid attention to when qutebrowser initially grew (I didn't know better at the time). At some point I should probably sit down with a piece of paper and find a suitable split between modules which doesn't require low-level modules to import things from higher-level ones. But that will be a bigger refactoring.

A different solution would be to move `from qutebrowser.misc import guiprocess` to `_execute_fileselect_command` in `browser.shared`. No hard feelings either way, though.

Reply:
Can be tricky with these cyclic imports I find. For now I left the import as is but changed the comment above it.